### Introduction

This project examines the tweet archive for WeRateDogs to draw insights from the dataset regarding the ratings of dogs. The Gather, Assess and Clean process is followed with the aim of answering the following questions:

- What type of dogs have the highest ratings?
- What type of dogs are the most popular in terms of retweet count and favorite count?
- What dog stage has the highest ratings?

## Gather

In [1]:
import json
import numpy as np
import os
import pandas as pd
import re
import requests
import tweepy
from timeit import default_timer as timer
from tweepy import OAuthHandler

In [2]:
# Load twitter archive file
archive_enhanced = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
# Download image-predictions file
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)
with open('image_predictions.tsv', mode='wb') as file:
    file.write(response.content)

In [4]:
# Import image predictions file
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')

In [5]:
# Query Twitter API for each tweet in the Twitter archive and save JSON in a text file
consumer_key = 'HIDDEN'
consumer_secret = 'HIDDEN'
access_token = 'HIDDEN'
access_secret = 'HIDDEN'

auth = OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit=True, wait_on_rate_limit_notify=True)

In [6]:
# Tweet IDs for which to gather additional data via Twitter's API
tweet_ids = archive_enhanced.tweet_id.values
len(tweet_ids)

2356

In [7]:
# Query Twitter's API for JSON data for each tweet ID in the Twitter archive
if not os.path.exists('tweet_json.txt'):
    count = 0
    fails_dict = {}
    start = timer()
    # Save each tweet's returned JSON as a new line in a .txt file
    with open('tweet_json.txt', 'w') as outfile:
        # This loop will likely take 20-30 minutes to run because of Twitter's rate limit
        for tweet_id in tweet_ids:
            count += 1
            print(str(count) + ": " + str(tweet_id))
            try:
                tweet = api.get_status(tweet_id, tweet_mode='extended')
                print("Success")
                json.dump(tweet._json, outfile)
                outfile.write('\n')
            except tweepy.TweepError as e:
                print("Fail")
                fails_dict[tweet_id] = e
                pass
    end = timer()
    print(end - start)
    print(fails_dict)


In [8]:
# DELETE
# Temp - save fails_dict to file for working with
'''
import csv
fails_list = []
for key in fails_dict.keys():
    fails_list.append(key)
fails_list
with open('fails_dict.csv', 'w') as outfile:
    wr = csv.writer(outfile, dialect='excel')
    wr.writerow(fails_list)
'''

"\nimport csv\nfails_list = []\nfor key in fails_dict.keys():\n    fails_list.append(key)\nfails_list\nwith open('fails_dict.csv', 'w') as outfile:\n    wr = csv.writer(outfile, dialect='excel')\n    wr.writerow(fails_list)\n"

In [9]:
# DELETE
# Working - uncomment out if necessary
'''
failures = []
with open('fails_dict.csv', 'r') as f:
    reader = csv.reader(f)
    for line in reader:
        failures.append(line)
'''

"\nfailures = []\nwith open('fails_dict.csv', 'r') as f:\n    reader = csv.reader(f)\n    for line in reader:\n        failures.append(line)\n"

In [10]:
# Load extended tweet data
tweets = []
with open('tweet_json.txt') as json_file:
    for line in json_file:
        tweets.append(json.loads(line))

# Place extended tweet data into a dataframe
extended_data = pd.DataFrame(tweets)

## Assess

In [11]:
# View column types and missing data in twitter_archive_enhanced
archive_enhanced.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [12]:
# Visually assess data in archive
archive_enhanced.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1255  710609963652087808                    NaN                  NaN   
135   866450705531457537                    NaN                  NaN   
44    883838122936631299                    NaN                  NaN   
1707  680801747103793152                    NaN                  NaN   
2268  667517642048163840                    NaN                  NaN   

                      timestamp  \
1255  2016-03-17 23:33:12 +0000   
135   2017-05-22 00:28:40 +0000   
44    2017-07-09 00:00:04 +0000   
1707  2015-12-26 17:25:59 +0000   
2268  2015-11-20 01:39:42 +0000   

                                                 source  \
1255  <a href="http://vine.co" rel="nofollow">Vine -...   
135   <a href="http://twitter.com/download/iphone" r...   
44    <a href="http://twitter.com/download/iphone" r...   
1707  <a href="http://twitter.com/download/iphone" r...   
2268  <a href="http://twitter.com" rel="nofollow">Tw...   

                                                   text  retweeted_status_id  \
1255  I've watched this a million times and you prob...                  NaN   
135   This is Jamesy. He gives a kiss to every other...                  NaN   
44    This is Noah. He can't believe someone made th...                  NaN   
1707  Great picture here. Dog on the right panicked ...                  NaN   
2268  This is Dook &amp; Milo. Dook is struggling to...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1255                       NaN                        NaN   
135                        NaN                        NaN   
44                         NaN                        NaN   
1707                       NaN                        NaN   
2268                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1255                      https://vine.co/v/idaTpwH5TgU                12   
135   https://twitter.com/dog_rates/status/866450705...                13   
44    https://twitter.com/dog_rates/status/883838122...                12   
1707  https://twitter.com/dog_rates/status/680801747...                10   
2268  https://twitter.com/dog_rates/status/667517642...                 8   

      rating_denominator    name doggo floofer  pupper puppo  
1255                  10    None  None    None    None  None  
135                   10  Jamesy  None    None  pupper  None  
44                    10    Noah  None    None    None  None  
1707                  10    None  None    None    None  None  
2268                  10    Dook  None    None    None  None

In [13]:
# Check for duplicated tweets
len(archive_enhanced[archive_enhanced.tweet_id.duplicated()])

0

In [14]:
# Check for inclusion of retweets
sum(~archive_enhanced.retweeted_status_id.isnull())

181

There are 181 tweets that are retweets and need to be removed.

In [15]:
# Check values for ratings_numerator
archive_enhanced.rating_numerator.value_counts().sort_index()

0         2
1         9
2         9
3        19
4        17
5        37
6        32
7        55
8       102
9       158
10      461
11      464
12      558
13      351
14       54
15        2
17        1
20        1
24        1
26        1
27        1
44        1
45        1
50        1
60        1
75        2
80        1
84        1
88        1
99        1
121       1
143       1
144       1
165       1
182       1
204       1
420       2
666       1
960       1
1776      1
Name: rating_numerator, dtype: int64

All of the numerators are whole numbers, however, a visual insepction of the data shows that there are some fractional numerators in the dataset.

In [16]:
# Check values in rating_denominator - should be 10
archive_enhanced.rating_denominator.value_counts().sort_index()

0         1
2         1
7         1
10     2333
11        3
15        1
16        1
20        2
40        1
50        3
70        1
80        2
90        1
110       1
120       1
130       1
150       1
170       1
Name: rating_denominator, dtype: int64

Not all of the denominators are 10, however, the requirement to have 10 as a denominator is not hard so these can be allowed.

In [17]:
# Examine data types and missing data in image_predictions
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


No data is missing as each column has the full number of entries (2075).

In [18]:
# Visually examine image_prdictions data
image_predictions.sample(10)

tweet_id                                          jpg_url  \
34    666435652385423360  https://pbs.twimg.com/media/CT-nsTQWEAEkyDn.jpg   
80    667455448082227200  https://pbs.twimg.com/media/CUNHMXTU8AAS3HH.jpg   
732   686730991906516992  https://pbs.twimg.com/media/CYfCMdFWAAA44YA.jpg   
1778  828408677031882754  https://pbs.twimg.com/media/C38ZSzlWIAEpQzs.jpg   
931   703079050210877440  https://pbs.twimg.com/media/CcHWqQCW8AEb0ZH.jpg   
1487  782021823840026624  https://pbs.twimg.com/media/CdHwZd0VIAA4792.jpg   
790   690649993829576704  https://pbs.twimg.com/media/CZWugJsWYAIzVzJ.jpg   
668   683030066213818368  https://pbs.twimg.com/media/CXqcOHCUQAAugTB.jpg   
1490  782722598790725632  https://pbs.twimg.com/media/CtzKC7zXEAALfSo.jpg   
981   707377100785885184  https://pbs.twimg.com/media/CdEbt0NXIAQH3Aa.jpg   

      img_num                        p1   p1_conf  p1_dog  \
34          1  Chesapeake_Bay_retriever  0.184130    True   
80          1           Tibetan_terrier  0.676376    True   
732         1           Tibetan_mastiff  0.338812    True   
1778        1                Weimaraner  0.133033    True   
931         2                  Pembroke  0.778503    True   
1487        1          golden_retriever  0.383223    True   
790         1                   bighorn  0.215438   False   
668         1                     boxer  0.722218    True   
1490        1              Irish_setter  0.574557    True   
981         1          golden_retriever  0.637225    True   

                            p2   p2_conf  p2_dog  \
34                   chain_saw  0.056775   False   
80               Irish_terrier  0.054933    True   
732               Newfoundland  0.180925    True   
1778  Chesapeake_Bay_retriever  0.092227    True   
931          Shetland_sheepdog  0.093834    True   
1487            cocker_spaniel  0.165930    True   
790                      hyena  0.137928   False   
668               bull_mastiff  0.193804    True   
1490          golden_retriever  0.339251    True   
981                 bloodhound  0.094542    True   

                                  p3   p3_conf  p3_dog  
34                       power_drill  0.036763   False  
80                 Yorkshire_terrier  0.040576    True  
732                 golden_retriever  0.180023    True  
1778  American_Staffordshire_terrier  0.065094    True  
931                         Cardigan  0.060296    True  
1487        Chesapeake_Bay_retriever  0.118199    True  
790                 Mexican_hairless  0.098171    True  
668                   French_bulldog  0.055194    True  
1490                       seat_belt  0.046108   False  
981                   cocker_spaniel  0.069797    True

The px columns are inconsistent in their string formatting and the table is untidy as there are multiple predictions in each row.

In [19]:
# Check column types and missing values in extended_data
extended_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2339 entries, 0 to 2338
Data columns (total 32 columns):
contributors                     0 non-null object
coordinates                      0 non-null object
created_at                       2339 non-null object
display_text_range               2339 non-null object
entities                         2339 non-null object
extended_entities                2065 non-null object
favorite_count                   2339 non-null int64
favorited                        2339 non-null bool
full_text                        2339 non-null object
geo                              0 non-null object
id                               2339 non-null int64
id_str                           2339 non-null object
in_reply_to_screen_name          77 non-null object
in_reply_to_status_id            77 non-null float64
in_reply_to_status_id_str        77 non-null object
in_reply_to_user_id              77 non-null float64
in_reply_to_user_id_str          77 non-null obj

In [20]:
# Visually examine extended_data
extended_data.head(5)

contributors coordinates                      created_at display_text_range  \
0         None        None  Tue Aug 01 16:23:56 +0000 2017            [0, 85]   
1         None        None  Tue Aug 01 00:17:27 +0000 2017           [0, 138]   
2         None        None  Mon Jul 31 00:18:03 +0000 2017           [0, 121]   
3         None        None  Sun Jul 30 15:58:51 +0000 2017            [0, 79]   
4         None        None  Sat Jul 29 16:00:24 +0000 2017           [0, 138]   

                                            entities  \
0  {'hashtags': [], 'symbols': [], 'user_mentions...   
1  {'hashtags': [], 'symbols': [], 'user_mentions...   
2  {'hashtags': [], 'symbols': [], 'user_mentions...   
3  {'hashtags': [], 'symbols': [], 'user_mentions...   
4  {'hashtags': [{'text': 'BarkWeek', 'indices': ...   

                                   extended_entities  favorite_count  \
0  {'media': [{'id': 892420639486877696, 'id_str'...           37783   
1  {'media': [{'id': 892177413194625024, 'id_str'...           32453   
2  {'media': [{'id': 891815175371796480, 'id_str'...           24437   
3  {'media': [{'id': 891689552724799489, 'id_str'...           41113   
4  {'media': [{'id': 891327551943041024, 'id_str'...           39326   

   favorited                                          full_text   geo  ...  \
0      False  This is Phineas. He's a mystical boy. Only eve...  None  ...   
1      False  This is Tilly. She's just checking pup on you....  None  ...   
2      False  This is Archie. He is a rare Norwegian Pouncin...  None  ...   
3      False  This is Darla. She commenced a snooze mid meal...  None  ...   
4      False  This is Franklin. He would like you to stop ca...  None  ...   

   quoted_status quoted_status_id quoted_status_id_str  \
0            NaN              NaN                  NaN   
1            NaN              NaN                  NaN   
2            NaN              NaN                  NaN   
3            NaN              NaN                  NaN   
4            NaN              NaN                  NaN   

   quoted_status_permalink retweet_count  retweeted retweeted_status  \
0                      NaN          8233      False              NaN   
1                      NaN          6084      False              NaN   
2                      NaN          4027      False              NaN   
3                      NaN          8384      False              NaN   
4                      NaN          9089      False              NaN   

                                              source truncated  \
0  <a href="http://twitter.com/download/iphone" r...     False   
1  <a href="http://twitter.com/download/iphone" r...     False   
2  <a href="http://twitter.com/download/iphone" r...     False   
3  <a href="http://twitter.com/download/iphone" r...     False   
4  <a href="http://twitter.com/download/iphone" r...     False   

                                                user  
0  {'id': 4196983835, 'id_str': '4196983835', 'na...  
1  {'id': 4196983835, 'id_str': '4196983835', 'na...  
2  {'id': 4196983835, 'id_str': '4196983835', 'na...  
3  {'id': 4196983835, 'id_str': '4196983835', 'na...  
4  {'id': 4196983835, 'id_str': '4196983835', 'na...  

[5 rows x 32 columns]

Most of the columns will not be required so there isn't much cleaning required here other than the id column which will need to be converted to a string. The id_str column won't be used as it won't match the tweet_id column in the archive_enhanced data, which is currently an int.

In [21]:
# Find number of retweets in extended_data
len(extended_data[~extended_data['retweeted_status'].isnull()])

167

There are 167 tweets that are retweets and will need to be removed from the data. These will be cleaned in the merged archive_enhanced data.

In [22]:
# Check for duplication between archive_enhanced and image_predictions
all_columns = pd.Series(list(archive_enhanced) + list(image_predictions))
all_columns[all_columns.duplicated()]

17    tweet_id
dtype: object

Only the tweet_id column was duplicated, which we can use as the link between to two tables.

#### Quality
##### `archive_enhanced` table
- data missing such as retweet_count and favorite_count
- tweet_id is an int (should be str)
- presence of retweets (retweeted_status_id is non-null)
- rating_numerator and rating_denominator columns are ints (should be floats)
- several incorrect ratings to be corrected

##### `image_predictions` table
- tweet_id is an int (should be str)
- dog names in p1, p2, p3 columns inconsistent - upper and lowercase names, inclusion of underscores etc.
- predictions include non-dog objects which should be removed prior to analysis

##### `extended_data` table
- presence of retweets (retweeted_status is non-null) (Note: will be cleaned at the same time that the archive_enhanced file is cleaned, as the two dataframes will have been merged at that point).

#### Tidiness
- `archive_enhanced` contains multiple columns for the dog stage which should be one column with the dog stage as the variable
- `image_predictions` has multiple predictions per row

## Clean

In [23]:
# Make copies of each dataframe for cleaning
archive_enhanced_clean = archive_enhanced.copy()
image_predictions_clean = image_predictions.copy()
extended_data_clean = extended_data.copy()

### Missing Data

#### `extended_info`: favorite_count and retweet_count need to be joined to `archive_enhanced`.

##### Define
- Drop the unwanted columns from `extended_data_clean` so just id, favorite_count and retweet_count remain
- Rename id to tweet_id in `extended_data_clean`
- Merge `extended_data_clean` with `archive_enhanced_clean` on the tweet_id column, keeping only the rows (tweet_id) that are present in both dataframes

##### Code

In [24]:
# Remove unwanted columns
columns = ['id', 'favorite_count', 'retweet_count']
extended_data_clean = extended_data_clean[columns]

# Rename id column
extended_data_clean = extended_data_clean.rename(columns={'id':'tweet_id'})
extended_data_clean.head()

tweet_id  favorite_count  retweet_count
0  892420643555336193           37783           8233
1  892177421306343426           32453           6084
2  891815181378084864           24437           4027
3  891689557279858688           41113           8384
4  891327558926688256           39326           9089

In [25]:
# Merge archive_enhanced_clean and extended_data_clean
archive_enhanced_clean = pd.merge(archive_enhanced_clean, \
                                  extended_data_clean, on='tweet_id', how='inner')

##### Test

In [26]:
# Check that retweet_count and favorite_count columns have been added and there are 2339 rows
archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2339 entries, 0 to 2338
Data columns (total 19 columns):
tweet_id                      2339 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2339 non-null object
source                        2339 non-null object
text                          2339 non-null object
retweeted_status_id           167 non-null float64
retweeted_status_user_id      167 non-null float64
retweeted_status_timestamp    167 non-null object
expanded_urls                 2280 non-null object
rating_numerator              2339 non-null int64
rating_denominator            2339 non-null int64
name                          2339 non-null object
doggo                         2339 non-null object
floofer                       2339 non-null object
pupper                        2339 non-null object
puppo                         2339 non-null object
favorite_count                23

### Tidiness

#### `archive_enhanced` contains multiple columns for the dog stage.

##### Define
- Create a column in `archive_enhanced` called dog_stage that holds the concatenation of each value in the doggo, floofer, pupper and puppo columns.
- Extract the correct dog stage from the dog_stage column and update
- For entries with more than one correct dog, record these as 'mixed'

##### Code

In [27]:
# Add dog_stage column and populate with concatenated values for dog stages
archive_enhanced_clean['dog_stage'] = archive_enhanced_clean['doggo'] \
    + archive_enhanced_clean['floofer'] + archive_enhanced_clean['pupper'] \
    + archive_enhanced_clean['puppo']

In [28]:
# Check values present in dog_stage column
archive_enhanced_clean['dog_stage'].value_counts()

NoneNoneNoneNone        1961
NoneNonepupperNone       244
doggoNoneNoneNone         82
NoneNoneNonepuppo         29
doggoNonepupperNone       12
NoneflooferNoneNone        9
doggoNoneNonepuppo         1
doggoflooferNoneNone       1
Name: dog_stage, dtype: int64

In [29]:
# Function to extract dog_stage
def get_dog_stage(current):
    '''Extract dog_stage from passed string'''
    if current == 'NoneNoneNoneNone':
        return 'None'
    elif current == 'doggoNoneNoneNone':
        return 'doggo'
    elif current == 'NoneflooferNoneNone':
        return 'floofer'
    elif current == 'NoneNonepupperNone':
        return 'pupper'
    elif current == 'NoneNoneNonepuppo':
        return 'puppo'
    else:
        return 'mixed'

In [30]:
# Update dog_stage column with call to get_dog_stage()
archive_enhanced_clean['dog_stage'] = archive_enhanced_clean['dog_stage'].apply(get_dog_stage)
# Drop individual dog style columns
archive_enhanced_clean = archive_enhanced_clean.drop(['doggo', 'puppo', 'pupper', \
                                                      'floofer'], axis=1)

##### Test

In [31]:
# Confirm that column has been updated correctly
archive_enhanced_clean['dog_stage'].value_counts()

None       1961
pupper      244
doggo        82
puppo        29
mixed        14
floofer       9
Name: dog_stage, dtype: int64

In [32]:
# Confirm that old dog stage columns have been dropped
archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2339 entries, 0 to 2338
Data columns (total 16 columns):
tweet_id                      2339 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2339 non-null object
source                        2339 non-null object
text                          2339 non-null object
retweeted_status_id           167 non-null float64
retweeted_status_user_id      167 non-null float64
retweeted_status_timestamp    167 non-null object
expanded_urls                 2280 non-null object
rating_numerator              2339 non-null int64
rating_denominator            2339 non-null int64
name                          2339 non-null object
favorite_count                2339 non-null int64
retweet_count                 2339 non-null int64
dog_stage                     2339 non-null object
dtypes: float64(4), int64(5), object(7)
memory usage: 310.6+ KB


#### `image_predictions` has multiple predictions per row.

##### Define
- Create separate dataframes from `image_predictions_clean` for p1, p2 and p3 data and keep only tweet_id, img_num, and the columns for that probability (`p1_df`, `p2_df`, `p3_df`)
- For each new dataframe, create a column that holds the p_order value (e.g. 1 for p1)
- Rename p1 column to 'p_type'
- Rename p1-conf to 'p_confidence'
- Rename p1_dog to 'p_dog'
- Repeat for p2 and p3
- Append `p2` to `p1`
- Append `p3` to `p1`
- Store `p1` as `image_predictions_clean`

##### Code

In [33]:
# Create new dataframes
p1_df = image_predictions_clean[['tweet_id', 'img_num', 'p1', 'p1_conf', 'p1_dog']]
p2_df = image_predictions_clean[['tweet_id', 'img_num', 'p2', 'p2_conf', 'p2_dog']]
p3_df = image_predictions_clean[['tweet_id', 'img_num', 'p3', 'p3_conf', 'p3_dog']]

# Rename p columns so consistent
p1_df = p1_df.rename(columns={'p1':'p_type', 'p1_conf':'p_conf', 'p1_dog':'p_dog'})
p2_df = p2_df.rename(columns={'p2':'p_type', 'p2_conf':'p_conf', 'p2_dog':'p_dog'})
p3_df = p3_df.rename(columns={'p3':'p_type', 'p3_conf':'p_conf', 'p3_dog':'p_dog'})

# Append tables and then assign to image_predictions_clean
p1_df = p1_df.append(p2_df)
p1_df = p1_df.append(p3_df)
image_predictions_clean = p1_df

##### Test

In [34]:
# Test dataframe contains correct columns (tweet_id, img_num, p_type, p_conf, p_dog)
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6225 entries, 0 to 2074
Data columns (total 5 columns):
tweet_id    6225 non-null int64
img_num     6225 non-null int64
p_type      6225 non-null object
p_conf      6225 non-null float64
p_dog       6225 non-null bool
dtypes: bool(1), float64(1), int64(2), object(1)
memory usage: 249.2+ KB


In [35]:
# Test data correct
image_predictions_clean.sample(10)

tweet_id  img_num                          p_type    p_conf  \
613   680145970311643136        1                      toy_poodle  0.226481   
1164  735256018284875776        1                         doormat  0.028075   
688   684177701129875456        1                 German_shepherd  0.162647   
1843  838561493054533637        1                       dalmatian  0.132820   
768   689280876073582592        3                       Chihuahua  0.637546   
768   689280876073582592        3  American_Staffordshire_terrier  0.150694   
258   670780561024270336        1             American_black_bear  0.056994   
288   671159727754231808        1                         pitcher  0.117446   
748   687807801670897665        1       Staffordshire_bullterrier  0.151113   
576   678675843183484930        1                            maze  0.339850   

      p_dog  
613    True  
1164  False  
688    True  
1843   True  
768    True  
768    True  
258   False  
288   False  
748    True  
576   False

### Quality

#### `archive_enhanced`: tweet_id is an int

##### Define
- Convert tweet_id column to str using .astype()

##### Code

In [36]:
archive_enhanced_clean.tweet_id = archive_enhanced_clean.tweet_id.astype(str)

##### Test

In [37]:
# Check that tweet_id is an object ('O')
archive_enhanced_clean.tweet_id.dtype

dtype('O')

#### `image_predictions`: tweet_id is an int

##### Define
- Convert tweet_id column to str using .astype()

##### Code

In [38]:
image_predictions_clean.tweet_id = image_predictions_clean.tweet_id.astype(str)

##### Test

In [39]:
# Check that tweet_id is an object ('O')
image_predictions_clean.tweet_id.dtype

dtype('O')

#### `archive_enhanced`: presence of retweets

##### Define
- Remove rows where retweeted_status_id != null (presence of a value means that the row is a retweet)

##### Code

In [40]:
# Check the number of rows in archive_enhanced_clean
archive_enhanced_clean.shape

(2339, 16)

In [41]:
# Check the number of null rows in retweeted_status_id (i.e. non-retweets) 
sum(archive_enhanced_clean['retweeted_status_id'].isnull())

2172

In [42]:
# Remove non-null rows (retweeted_status_id) from archive_enhanced
archive_enhanced_clean = archive_enhanced_clean[archive_enhanced_clean[
    'retweeted_status_id'].isnull()]

##### Test

In [43]:
# Check that number of rows = 2172
archive_enhanced_clean.shape

(2172, 16)

In [44]:
# Check that there are no rows with a non-null value in the retweeted_satus_id column
sum(~archive_enhanced_clean['retweeted_status_id'].isnull())

0

#### rating_numerator and rating_denominator columns are ints

##### Define
- Convert rating_denominator column to float using .astype()

##### Code

In [45]:
# Convert each column to an int
archive_enhanced_clean.rating_denominator = archive_enhanced_clean.rating_denominator.astype(float)

Note that the rating_numerator column will be addressed when the regex for extracting the numerator from the text column is updated.

##### Test

In [46]:
# Test that rating_denomiator column type has changed to float
archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2172 entries, 0 to 2338
Data columns (total 16 columns):
tweet_id                      2172 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2172 non-null object
source                        2172 non-null object
text                          2172 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2114 non-null object
rating_numerator              2172 non-null int64
rating_denominator            2172 non-null float64
name                          2172 non-null object
favorite_count                2172 non-null int64
retweet_count                 2172 non-null int64
dog_stage                     2172 non-null object
dtypes: float64(5), int64(3), object(8)
memory usage: 288.5+ KB


#### Incorrect rating_numerator values

##### Define
- Write a regex to find the numerator from the text column
- Save numerator to rating_numerator column
- Convert rating_numerator clumn to float using .astype()

##### Code

In [47]:
# Get value counts for ratings_numerator
archive_enhanced_clean.rating_numerator.value_counts().sort_index()

0         2
1         8
2         9
3        19
4        17
5        36
6        32
7        54
8        98
9       156
10      441
11      426
12      499
13      306
14       43
15        1
17        1
20        1
24        1
26        1
27        1
44        1
45        1
50        1
60        1
75        1
80        1
84        1
88        1
99        1
121       1
143       1
144       1
165       1
182       1
204       1
420       2
666       1
960       1
1776      1
Name: rating_numerator, dtype: int64

In [48]:
# Use a regex to include the fraction for the numerator
correct_numerator = archive_enhanced_clean.text.str.extract('((?:\d+\.)?\d+)\/(\d+)', expand=True)
archive_enhanced_clean['rating_numerator'] = correct_numerator

# Convert the rating_numerator column back to a float
archive_enhanced_clean['rating_numerator'] = archive_enhanced_clean['rating_numerator'].astype(float)

##### Test

In [49]:
# Check values in rating_numerator column now include fractional numerators
archive_enhanced_clean.rating_numerator.value_counts().sort_index()

0.00         2
1.00         8
2.00         9
3.00        19
4.00        17
5.00        34
6.00        32
7.00        54
8.00        98
9.00       156
9.50         1
9.75         1
10.00      441
11.00      426
11.26        1
11.27        1
12.00      499
13.00      306
13.50        1
14.00       43
15.00        1
17.00        1
20.00        1
24.00        1
44.00        1
45.00        1
50.00        1
60.00        1
80.00        1
84.00        1
88.00        1
99.00        1
121.00       1
143.00       1
144.00       1
165.00       1
182.00       1
204.00       1
420.00       2
666.00       1
960.00       1
1776.00      1
Name: rating_numerator, dtype: int64

We can see that there are now some fractional numerators included such as 9.5, 9.75 and 11.26

In [50]:
# Check data type for rating_numerator column
archive_enhanced_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2172 entries, 0 to 2338
Data columns (total 16 columns):
tweet_id                      2172 non-null object
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2172 non-null object
source                        2172 non-null object
text                          2172 non-null object
retweeted_status_id           0 non-null float64
retweeted_status_user_id      0 non-null float64
retweeted_status_timestamp    0 non-null object
expanded_urls                 2114 non-null object
rating_numerator              2172 non-null float64
rating_denominator            2172 non-null float64
name                          2172 non-null object
favorite_count                2172 non-null int64
retweet_count                 2172 non-null int64
dog_stage                     2172 non-null object
dtypes: float64(6), int64(2), object(8)
memory usage: 288.5+ KB


The rating_numerator column is now a float.

#### Inconsistent format for dog names in p columns of `image_predictions`.

##### Define
- Replace _ with ' ' in p_type column of `image_predictions_clean`
- Capitalise each word in p_type column of `image_predictions_clean`

##### Code

In [51]:
# Replace '_'
image_predictions_clean.p_type = image_predictions_clean.p_type.str.replace("_", " ")

# Capitalize first letters of each word
image_predictions_clean.p_type = image_predictions_clean.p_type.str.title()

##### Test

In [52]:
image_predictions_clean.sample(15)

tweet_id  img_num                       p_type    p_conf  \
1135  728751179681943552        1                       Borzoi  0.037976   
1641  807106840509214720        1                   Pomeranian  0.120358   
1029  711008018775851008        1               French Bulldog  0.731405   
1342  758854675097526272        4                     Komondor  0.001246   
760   688828561667567616        1                        Skunk  0.139392   
583   678800283649069056        1                     Airedale  0.122701   
152   668645506898350081        1                 Balance Beam  0.084076   
1103  721503162398597120        3                    Chihuahua  0.001248   
620   680473011644985345        1  West Highland White Terrier  0.138709   
1430  773191612633579521        1             Blenheim Spaniel  0.427766   
1555  793195938047070209        2           Labrador Retriever  0.654762   
1295  751937170840121344        1             Lakeland Terrier  0.424168   
465   675006312288268288        1                 Space Heater  0.043389   
586   679047485189439488        1                    Bannister  0.025248   
1768  827199976799354881        4                   Great Dane  0.869681   

      p_dog  
1135   True  
1641   True  
1029   True  
1342   True  
760   False  
583    True  
152   False  
1103   True  
620    True  
1430   True  
1555   True  
1295   True  
465   False  
586   False  
1768   True

#### Predictions include non-dog objects in `image_predictions`.

##### Define
- Drop rows where p_dog == False in `image_predictions_clean`

##### Code

In [53]:
# Get initial number of rows in image_predictions_clean
image_predictions_clean.shape

(6225, 5)

In [54]:
# Keep only rows where 'p_dog' == True
image_predictions_clean = image_predictions_clean[image_predictions_clean['p_dog'] == True]

##### Test

In [55]:
# Test that rows have been dropped
image_predictions_clean.shape

(4584, 5)

In [56]:
# Test that there are no False in 'p_dog' column
image_predictions_clean.p_dog.value_counts()

True    4584
Name: p_dog, dtype: int64

## Storage

In [57]:
archive_enhanced_clean.to_csv('twitter_archive_master.csv', index=False)
image_predictions_clean.to_csv('image_predictions_master.csv', index=False)

## Analysis

#### Question 1: Which types of dogs have the highest ratings?

To answer this question the ratings for each tweet need to be calculated by dividing the rating_numerator column by the rating_denominator column.

In [58]:
# Create a column that holds the rating (rating_numerator / rating_denominator)
archive_enhanced_clean['rating'] = archive_enhanced_clean['rating_numerator'] \
                                    / archive_enhanced_clean['rating_denominator']

For the dog types at least a confidence level of 50% for the dog type will be required, so tweets with an image confidence below this will need to be removed prior to joining the tables.

In [59]:
# Remove rows with a p_conf < 0.5
image_predictions_clean = image_predictions_clean.query('p_conf >= 0.5')
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 973 entries, 1 to 2072
Data columns (total 5 columns):
tweet_id    973 non-null object
img_num     973 non-null int64
p_type      973 non-null object
p_conf      973 non-null float64
p_dog       973 non-null bool
dtypes: bool(1), float64(1), int64(1), object(2)
memory usage: 39.0+ KB


In [60]:
# Check that to tweets are duplicated (having two dog type predictions remaining in dataset)
sum(image_predictions_clean.tweet_id.duplicated())

0

In [61]:
# Create a table that holds the archive_enhanced data joined with image_predictions
archive_images = pd.merge(image_predictions_clean, archive_enhanced_clean, on='tweet_id', how='inner')

In [65]:
archive_images

tweet_id  img_num                       p_type    p_conf  \
0    666029285002620928        1                      Redbone  0.506826   
1    666033412701032449        1              German Shepherd  0.596461   
2    666049248165822465        1           Miniature Pinscher  0.560311   
3    666050758794694657        1         Bernese Mountain Dog  0.651137   
4    666055525042405380        1                         Chow  0.692517   
5    666063827256086533        1             Golden Retriever  0.775930   
6    666071193221509120        1                Gordon Setter  0.503672   
7    666099513787052032        1                        Lhasa  0.582330   
8    666287406224695296        1                  Maltese Dog  0.857531   
9    666345417576210432        1             Golden Retriever  0.858744   
10   666396247373291520        1                    Chihuahua  0.978108   
11   666407126856765440        1      Black-And-Tan Coonhound  0.529139   
12   666421158376562688        1             Blenheim Spaniel  0.906777   
13   666437273139982337        1                    Chihuahua  0.671853   
14   666691418707132416        1              German Shepherd  0.975401   
15   666701168228331520        1           Labrador Retriever  0.887707   
16   666739327293083650        1             Miniature Poodle  0.546933   
17   666781792255496192        1            Italian Greyhound  0.618316   
18   667044094246576128        1             Golden Retriever  0.765266   
19   667062181243039745        1     Chesapeake Bay Retriever  0.825678   
20   667090893657276420        1                    Chihuahua  0.959514   
21   667119796878725120        1                     Pembroke  0.741563   
22   667138269671505920        1  West Highland White Terrier  0.747713   
23   667152164079423490        1                   Toy Poodle  0.535411   
24   667171260800061440        1              Giant Schnauzer  0.841265   
25   667182792070062081        1             Golden Retriever  0.949892   
26   667200525029539841        1               Siberian Husky  0.694904   
27   667393430834667520        1                     Papillon  0.557009   
28   667435689202614272        1                   Rottweiler  0.999091   
29   667453023279554560        1           Labrador Retriever  0.825670   
..                  ...      ...                          ...       ...   
904  882627270321602560        1                     Pembroke  0.542982   
905  882762694511734784        1           Labrador Retriever  0.850050   
906  883117836046086144        2             Golden Retriever  0.949562   
907  883360690899218434        1                         Chow  0.987997   
908  883482846933004288        1             Golden Retriever  0.943082   
909  883838122936631299        1                     Doberman  0.610946   
910  884162670584377345        1              German Shepherd  0.707046   
911  884441805382717440        1                     Pembroke  0.993225   
912  884562892145688576        1                          Pug  0.546406   
913  884876753390489601        1                         Chow  0.822103   
914  885167619883638784        4                     Malamute  0.812482   
915  885984800019947520        1             Blenheim Spaniel  0.972494   
916  886258384151887873        1                          Pug  0.943575   
917  886366144734445568        1               French Bulldog  0.999201   
918  886983233522544640        2                    Chihuahua  0.793469   
919  887101392804085760        1                      Samoyed  0.733942   
920  887473957103951883        2                     Pembroke  0.809197   
921  887705289381826560        1                       Basset  0.821664   
922  888078434458587136        1               French Bulldog  0.995026   
923  888554962724278272        3               Siberian Husky  0.700377   
924  888917238123831296        1             Golden Retriever  0.714719   
925  889278841981685760        1        

Dog types that have only a couple of tweets will likely skew the results, being far more likely to have a high mean etc. To counter this, only dog types with at least five tweets will be kept in the dataset to be analysed.

In [72]:
# Group tweets by predicted dog type and drop those with a count of less than 5
mask = archive_images.groupby('p_type').tweet_id.count()[lambda x: x >= 5]
new_archive = archive_images[mask]
new_archive

KeyError: "None of [Int64Index([  6,   9,   9,   8,  10,   6,   8,   6,   5,   8,   7,   5,  12,\n             11,  47,  26,  14,   5,   9,   7,   9,   7,  21,  19,   5, 116,\n              9,   5,  64,   6,  18,   5,  15,  11,   8,   5,   9,  70,  29,\n             43,  10,  29,   6,  12,  11,   8,   5,  10,  24,  10,  11,   6],\n           dtype='int64')] are in the [columns]"

In [69]:
# Check only dog types with at least 5 tweets are included and which are the most common
grouped_dog_type.sort_values('rating', ascending=False)

ValueError: No axis named rating for object type <class 'type'>

In [66]:
# Get the mean rating for each dog type
grouped_dog_type.rating.mean()

AttributeError: 'Series' object has no attribute 'rating'

The three highest rated dog types (by mean rating) are Afghan Hound, Curlt-Coated Retriever and Saluki. We need to check the number of each dog type in the data to understand the significance of these averages.

In [ ]:
# Check number of each dog type in sample
print('Afghan Hound: {}'.format(archive_images[archive_images['p_type'] == 'Afghan Hound'].p_type
                                .count()))
print('Curly-Coated Retriever: {}'.format(archive_images[archive_images['p_type'] == 
                                                         'Curly-Coated Retriever'].p_type.count()))
print('Saluki: {}'.format(archive_images[archive_images['p_type'] == 'Saluki'].p_type.count()))
print('Briard: {}'.format(archive_images[archive_images['p_type'] == 'Briard'].p_type.count()))
print('Tibetan Mastiff: {}'.format(archive_images[archive_images['p_type'] == 'Tibetan Mastiff']
                                   .p_type.count()))

The small number of samples for each dog type suggest that the average rating may not be that useful an indicator of popularity.

#### Question 2: What type of dogs are the most popular in terms of retweet count and favorite count?

In [ ]:
# Get the 5 highest retweet counts by dog type
grouped_dog_type.retweet_count.sum().sort_values(ascending=False).head()

The highest number of retweets were received by Golden Retrievers, Labrador Retrievers and Pembrokes.

In [ ]:
# Get the 5 highest retweet counts, by average, by dog type
grouped_dog_type.retweet_count.mean().sort_values(ascending=False).head()

When taken as an average, the highest retweet counts were received by Eskimo dogs, Saluki and Irish Water Spaniels.

In [ ]:
# Get the 5 highest favorite counts by dog type
grouped_dog_type.favorite_count.sum().sort_values(ascending=False).head()

The highest number of favorites were received by Golden Retrievers, Pembrokes andLabrador Retrievers.

In [ ]:
# Get the 5 highest favorite counts, by average, by dog type
grouped_dog_type.favorite_count.mean().sort_values(ascending=False).head()

When taken as an average, the highest favorite counts were received by Salukis, Eskimo dogs and Irish Water Spaniels.